juju/juju
# Replace environs.ReloadSpaces by network service

This patch removes the spaces functionality that was previously implemented in environs, thus relying only on domain services. The new service domain is now wired-up in the facades, and in the bootstrap worker. We therefore removed the reload spaces functionality from the agent bootstrap.

A small modification has been done on the Reload spaces at the service layer, where we now generate the UUID for the subnets that are to be inserted when doing subnet discovery.

Another fix that has been introduced in this PR is changing the primary key in `provider_network_subnet` and removing the `UNIQUE` index on `subnet_uuid`. This was an error found during the reload of spaces using lxd.

## Checklist



## QA steps

The main idea on this patch is to check the subnets that have been loaded from the underlying provider, therefore we can bootstrap on lxd and check the dqlite contents:
```
juju bootstrap localhost c
# connect to the controller machine:
juju ssh -m controller 0
# install and set-up for dqlite query:
sudo apt install go-dqlite
sudo snap install yq
sudo cat /var/lib/juju/agents/machine-0/agent.conf | yq '.controllercert' | xargs -I% echo % > dqlite.cert
sudo cat /var/lib/juju/agents/machine-0/agent.conf | yq '.controllerkey' | xargs -I% echo % > dqlite.key
sudo dqlite -s file:///var/lib/juju/dqlite/cluster.yaml -c ./dqlite.cert -k ./dqlite.key controller
# retrieve the controller model UUID:
dqlite> select * from model_list;
c5efa2e5-5c00-4685-8094-2c092788e629
# close dqlite and open a new cli pointing the controller model db:
sudo dqlite -s file:///var/lib/juju/dqlite/cluster.yaml -c ./dqlite.cert -k ./dqlite.key c5efa2e5-5c00-4685-8094-2c092788e629
dqlite> select * from subnet;
018ee7a7-de15-7192-95f0-0705e578aa3d|10.170.236.0/24|0|0|0
018ee7a7-de15-7195-ba6b-e3e6f92befea|10.68.58.0/24|0|0|0
018ee7a7-de15-719d-83f1-45c912fdb86e|192.168.49.0/24|0|0|0
018ee7a7-de15-71a0-8f95-622e60e9ad78|172.17.0.0/16|0|0|0
018ee7a7-de15-71a2-93d5-d2e469078aaa|172.18.0.0/16|0|0|0
018ee7a7-de15-71a4-81a2-110b129236de|10.215.127.0/24|0|0|0
```
You should be able to see the list of subnets on your local machine (compare with `ip a`).

For a simpler scenario, let's use use the spaces from the client directly:
```
juju add-model m
juju spaces
Name   Space ID  Subnets
alpha  0         10.170.236.0/24
                 10.215.127.0/24
                 10.68.58.0/24
                 172.17.0.0/16
                 172.18.0.0/16
                 192.168.49.0/24
# let's add a space on subnet `10.215.127.0/24`
juju add-space foo 10.68.58.0/24
juju spaces
Name   Space ID                              Subnets
alpha  0                                     10.170.236.0/24
                                             10.215.127.0/24
                                             172.17.0.0/16
                                             172.18.0.0/16
                                             192.168.49.0/24
foo    018f0628-f243-7934-9d72-14fbf4af6367  10.68.58.0/24
# now let's rename that space
juju rename-space foo bar
juju spaces
Name   Space ID                              Subnets
alpha  0                                     10.170.236.0/24
                                             10.215.127.0/24
                                             172.17.0.0/16
                                             172.18.0.0/16
                                             192.168.49.0/24
bar    018f0628-f243-7934-9d72-14fbf4af6367  10.68.58.0/24
```


## Links

**Jira card:** JUJU-5832
